src/OSX/Avalonia.MonoMac/ScreenImpl.cs

```diff
@@ -0,0 +1,62 @@
+﻿using Avalonia.Platform;
+using MonoMac.AppKit;
+using MonoMac.Foundation;
+
+namespace Avalonia.MonoMac
+{
+    public class ScreenImpl : IScreenImpl
+    {
+        private const string NSApplicationDidChangeScreenParametersNotification = "NSApplicationDidChangeScreenParametersNotification";
+     
+        public int ScreenCount
+        {
+            get => NSScreen.Screens.Length;
+        }
+
+        public Screen[] AllScreens
+        {
+            get
+            {
+                if (_allScreens == null)
+                {
+                    NSScreen[] screens = NSScreen.Screens;
+                    Screen[] s = new Screen[screens.Length];
+                    NSScreen primary = NSScreen.MainScreen;
+                    for (int i = 0; i < screens.Length; i++)
+                    {
+                        Rect bounds = screens[i].Frame.ToAvaloniaRect().ConvertRectY();
+                        Rect workArea = screens[i].VisibleFrame.ToAvaloniaRect().ConvertRectY();
+                        s[i] = new MacScreen(bounds, workArea, screens[i] == primary, screens[i].Handle);
+                    }
+
+                    _allScreens = s;
+                    _observer = NSNotificationCenter.DefaultCenter.AddObserver(NSApplicationDidChangeScreenParametersNotification, notification =>
+                                                                                                                                  {
+                                                                                                                                      _allScreens = null;
+                                                                                                                                      NSNotificationCenter
+                                                                                                                                          .DefaultCenter
+                                                                                                                                          .RemoveObserver(_observer);
+                                                                                                                                  });
+                }
+                return _allScreens;
+            }
+        }
+        
+        public Screen PrimaryScreen
```

Comment:
Since you are going through the array of cached instances, this logic should be moved to `Avalonia.Platform.Screen`

Reply:
Whoops, that shouldn't have been the cached instances